nbdime/merging/decisions.py

```diff
@@ -0,0 +1,462 @@
+    recursively, so a common chain of patches will be resolved as well.
+    """
+    assert isinstance(path, (tuple, list))
+    popped = _pop_path(diffs)
+    while popped:
+        path = path + (popped["key"],)
+        diffs = popped["diffs"]
+        popped = _pop_path(diffs)
+    return path, diffs
+
+
+def _pop_path(diffs):
+    """Pops of a common path from patch ops sharing a common key.
+
+    Checks whether all diffs are single patch operations sharing the same key,
+    or alternatively empty diffs. If so, it returns the shared path/key as well
+    as the inner diffs of the patch operations (in the same order as the passed
+    diffs).
+    """
+    key = None
+    popped_diffs = []
+    for d in diffs:
+        # Empty diffs can be skipped
+        if d is None or len(d) == 0:
+            popped_diffs.append(None)
+            continue
+        # Check that we have only one op, which is a patch op
+        if len(d) != 1 or d[0].op != DiffOp.PATCH:
+            return
+        # Ensure all present diffs have the same key
+        if key is None:
+            key = d[0].key
+        elif key != d[0].key:
+            return
+        # Ensure the sub diffs of all ops are suitable as outer layer
+        # if d[0].diff.length > 1:
+        #    return
+        popped_diffs.append(d[0].diff)
+    if key is None:
+        return
+    return {'key': key, 'diffs': popped_diffs}
+
+
+def push_path(path, diffs):
+    """Wraps the diffs in patch operations matching path.
+    """
+    for key in reversed(path):
+        diffs = [op_patch(key, diffs)]
+    return diffs
+
+
+def pop_patch_decision(decision):
+    """Create a new decision one level lower in diff tree.
+
+    Checks whether all diffs are single patch operations sharing the same key,
+    or alternatively empty diffs. A new decision is then created at that key.
+
+    Returns the new decision.
+
+    Raises a ValueErorr if a decision can not be created at the lower level.
+    """
+    diffs = [decision.local_diff, decision.remote_diff]
+    if decision.action == "custom":
+        diffs.append(decision.custom_diff)
+    popped = _pop_path(diffs)
+    if popped is None:
+        raise ValueError("Cannot pop patch decision for: " + str(decision))
+    ret = MergeDecision(
+        common_path=decision.common_path + (popped["key"],),
+        local_diff=popped["diffs"][0],
+        remote_diff=popped["diffs"][1],
+        action=decision.action,
+        conflict=decision.conflict)
+    if decision.action == "custom":
+        ret.custom_diff = popped["diffs"][2]
+    return ret
+
+
+def pop_all_patch_decisions(decision):
+    """Create a new decision at the furthest level in the diff tree.
+
+    Calls `pop_patch_decision` recursively until it is at the lowest possible
+    decision level.
+
+    If the decision is already at the lowest level, it returns the original
+    decision.
+    """
+    try:
+        while 1:
+            decision = pop_patch_decision(decision)
```

Comment:
Instead of `try/except` + `while 1:`, why not do `while decision is not None:`? (see above).